core/ordersync/ordersync.go

```diff
@@ -0,0 +1,507 @@
+		Metadata:    metadata,
+	}, nil
+}
+
+func parseResponseWithSubprotocol(subprotocol Subprotocol, providerID peer.ID, rawRes *rawResponse) (*Response, error) {
+	metadata, err := subprotocol.ParseResponseMetadata(rawRes.Metadata)
+	if err != nil {
+		return nil, err
+	}
+	return &Response{
+		ProviderID: providerID,
+		Orders:     rawRes.Orders,
+		Complete:   rawRes.Complete,
+		Metadata:   metadata,
+	}, nil
+}
+
+func (s *Service) getOrdersFromPeer(ctx context.Context, providerID peer.ID) error {
+	stream, err := s.node.NewStream(ctx, providerID, ID)
+	if err != nil {
+		s.handlePeerScoreEvent(providerID, psUnexpectedDisconnect)
+		return err
+	}
+	defer func() {
+		_ = stream.Close()
+	}()
+
+	var nextReq *Request
+	var selectedSubprotocol Subprotocol
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var rawReq *rawRequest
+		if nextReq == nil {
+			// First request
+			rawReq = &rawRequest{
+				Type:         TypeRequest,
+				Subprotocols: s.SupportedSubprotocols(),
+				Metadata:     nil,
+			}
+		} else {
+			encodedMetadata, err := json.Marshal(nextReq.Metadata)
+			if err != nil {
+				return err
+			}
+			rawReq = &rawRequest{
+				Type:         TypeRequest,
+				Subprotocols: []string{selectedSubprotocol.Name()},
+				Metadata:     encodedMetadata,
+			}
+		}
+
+		if err := json.NewEncoder(stream).Encode(rawReq); err != nil {
+			s.handlePeerScoreEvent(providerID, psUnexpectedDisconnect)
+			return err
+		}
+
+		rawRes, err := waitForResponse(ctx, stream)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, psValidMessage)
+
+		subprotocol, found := s.subprotocols[rawRes.Subprotocol]
+		if !found {
+			s.handlePeerScoreEvent(providerID, psSubprotocolNegotiationFailed)
+			return fmt.Errorf("unsupported subprotocol: %s", subprotocol)
+		}
+		selectedSubprotocol = subprotocol
+		res, err := parseResponseWithSubprotocol(subprotocol, providerID, rawRes)
+		if err != nil {
+			s.handlePeerScoreEvent(providerID, psInvalidMessage)
+			return err
+		}
+
+		nextReq, err = subprotocol.HandleOrderSyncResponse(ctx, res)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, receivedOrders)
+
+		if rawRes.Complete {
+			return nil
+		}
+	}
+}
+
+// shufflePeers randomizes the order of the given list of peers.
+func shufflePeers(peers []peer.ID) {
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(peers), func(i, j int) { peers[i], peers[j] = peers[j], peers[i] })
+}
+
+func waitForRequest(parentCtx context.Context, stream network.Stream) (*rawRequest, error) {
+	ctx, cancel := context.WithTimeout(parentCtx, requestResponseTimeout)
+	defer cancel()
+	reqChan := make(chan *rawRequest, 1)
+	go func() {
+		var rawReq rawRequest
+		if err := json.NewDecoder(stream).Decode(&rawReq); err != nil {
+			log.WithFields(log.Fields{
+				"error":     err.Error(),
+				"requester": stream.Conn().RemotePeer().Pretty(),
+			}).Warn("could not encode ordersync request")
```

Comment:
Doesn't `waitForRequest` want to return this error instead of sending it in the request channel? What am I missing?

Reply:
Good catch. We should be returning the error. Just fixed it.